In [2]:
# Import dependencies
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
import pandas as pd
import tensorflow as tf

#  Import and read the charity_data.csv.
import pandas as pd
application_df = pd.read_csv("https://static.bc-edx.com/data/dl-1-2/m21/lms/starter/charity_data.csv")
application_df.head()

EIN                                      NAME APPLICATION_TYPE  \
0  10520599              BLUE KNIGHTS MOTORCYCLE CLUB              T10   
1  10531628    AMERICAN CHESAPEAKE CLUB CHARITABLE TR               T3   
2  10547893        ST CLOUD PROFESSIONAL FIREFIGHTERS               T5   
3  10553066            SOUTHSIDE ATHLETIC ASSOCIATION               T3   
4  10556103  GENETIC RESEARCH INSTITUTE OF THE DESERT               T3   

        AFFILIATION CLASSIFICATION      USE_CASE  ORGANIZATION  STATUS  \
0       Independent          C1000    ProductDev   Association       1   
1       Independent          C2000  Preservation  Co-operative       1   
2  CompanySponsored          C3000    ProductDev   Association       1   
3  CompanySponsored          C2000  Preservation         Trust       1   
4       Independent          C1000     Heathcare         Trust       1   

      INCOME_AMT SPECIAL_CONSIDERATIONS  ASK_AMT  IS_SUCCESSFUL  
0              0                      N     5000              1  
1         1-9999                      N   108590              1  
2              0                      N     5000              0  
3    10000-24999                      N     6692              1  
4  100000-499999                      N   142590              1

In [3]:
# Drop the non-beneficial ID columns, 'EIN' and 'NAME'.
application_df = application_df.drop(columns=['EIN', 'NAME'])
application_df.head()

APPLICATION_TYPE       AFFILIATION CLASSIFICATION      USE_CASE  \
0              T10       Independent          C1000    ProductDev   
1               T3       Independent          C2000  Preservation   
2               T5  CompanySponsored          C3000    ProductDev   
3               T3  CompanySponsored          C2000  Preservation   
4               T3       Independent          C1000     Heathcare   

   ORGANIZATION  STATUS     INCOME_AMT SPECIAL_CONSIDERATIONS  ASK_AMT  \
0   Association       1              0                      N     5000   
1  Co-operative       1         1-9999                      N   108590   
2   Association       1              0                      N     5000   
3         Trust       1    10000-24999                      N     6692   
4         Trust       1  100000-499999                      N   142590   

   IS_SUCCESSFUL  
0              1  
1              1  
2              0  
3              1  
4              1

In [4]:
application_df.columns

Index(['APPLICATION_TYPE', 'AFFILIATION', 'CLASSIFICATION', 'USE_CASE',
       'ORGANIZATION', 'STATUS', 'INCOME_AMT', 'SPECIAL_CONSIDERATIONS',
       'ASK_AMT', 'IS_SUCCESSFUL'],
      dtype='object')

In [5]:
new_app_df= application_df[['AFFILIATION', 'USE_CASE', 'ORGANIZATION', 'STATUS', 'INCOME_AMT', 'SPECIAL_CONSIDERATIONS', 'ASK_AMT', 'IS_SUCCESSFUL']]
new_app_df.head()

AFFILIATION      USE_CASE  ORGANIZATION  STATUS     INCOME_AMT  \
0       Independent    ProductDev   Association       1              0   
1       Independent  Preservation  Co-operative       1         1-9999   
2  CompanySponsored    ProductDev   Association       1              0   
3  CompanySponsored  Preservation         Trust       1    10000-24999   
4       Independent     Heathcare         Trust       1  100000-499999   

  SPECIAL_CONSIDERATIONS  ASK_AMT  IS_SUCCESSFUL  
0                      N     5000              1  
1                      N   108590              1  
2                      N     5000              0  
3                      N     6692              1  
4                      N   142590              1

In [8]:
# Convert categorical data to numeric with `pd.get_dummies`
new_app_df_dummied = pd.get_dummies(new_app_df)
new_app_df_dummied

STATUS   ASK_AMT  IS_SUCCESSFUL  AFFILIATION_CompanySponsored  \
0           1      5000              1                             0   
1           1    108590              1                             0   
2           1      5000              0                             1   
3           1      6692              1                             1   
4           1    142590              1                             0   
...       ...       ...            ...                           ...   
34294       1      5000              0                             0   
34295       1      5000              0                             1   
34296       1      5000              0                             1   
34297       1      5000              1                             0   
34298       1  36500179              0                             0   

       AFFILIATION_Family/Parent  AFFILIATION_Independent  \
0                              0                        1   
1                              0                        1   
2                              0                        0   
3                              0                        0   
4                              0                        1   
...                          ...                      ...   
34294                          0                        1   
34295                          0                        0   
34296                          0                        0   
34297                          0                        1   
34298                          0                        1   

       AFFILIATION_National  AFFILIATION_Other  AFFILIATION_Regional  \
0                         0                  0                     0   
1                         0                  0                     0   
2                         0                  0                     0   
3                         0                  0                     0   
4                         0                  0                     0   
...                     ...                ...                   ...   
34294                     0                  0                     0   
34295                     0                  0                     0   
34296                     0                  0                     0   
34297                     0                  0                     0   
34298                     0                  0                     0   

       USE_CASE_CommunityServ  ...  INCOME_AMT_1-9999  INCOME_AMT_10000-24999  \
0                           0  ...                  0                       0   
1                           0  ...                  1                       0   
2                           0  ...                  0                       0   
3                           0  ...                  0                       1   
4                           0  ...                  0                       0   
...                       ...  ...                ...                     ...   
34294                       0  ...                  0                       0   
34295                       0  ...                  0                       0   
34296                       0  ...                  0                       0   
34297                       0  ...                  0                       0   
34298                       0  ...                  0                       0   

       INCOME_AMT_100000-499999  INCOME_AMT_10M-50M  INCOME_AMT_1M-5M  \
0                             0                   0                 0   
1                             0                   0                 0   
2                             0                   0                 0   
3                             0                   0                 0   
4                             1                   0                 0   
...                         ...                 ...               ...   
34294                         0                   0         

In [9]:
# Split our preprocessed data into our features and target arrays
X = new_app_df_dummied.drop(columns = 'IS_SUCCESSFUL')
y = new_app_df_dummied['IS_SUCCESSFUL']

# Split the preprocessed data into a training and testing dataset
X_train, X_test, y_train, y_test = train_test_split(X,y, random_state=1)

In [10]:
# Create a StandardScaler instances
scaler = StandardScaler()

# Fit the StandardScaler
X_scaler = scaler.fit(X_train)

# Scale the data
X_train_scaled = X_scaler.transform(X_train)
X_test_scaled = X_scaler.transform(X_test)

nn_optimization3 model

In [11]:
# removes CLASSIFICATION and APPLICATION_TYPE from the dataframe
# Define the model with the number of input features and hidden nodes for each layer.
number_input_features = X_train.shape[1]
hidden_nodes_layer1 =  80
hidden_nodes_layer2 = 30

nn_optimization3 = tf.keras.models.Sequential()

# First hidden layer
nn_optimization3.add(
    tf.keras.layers.Dense(units=hidden_nodes_layer1, input_dim=number_input_features, activation="relu")
)

# Second hidden layer
nn_optimization3.add(tf.keras.layers.Dense(units=hidden_nodes_layer2, activation="relu"))

# Output layer
nn_optimization3.add(tf.keras.layers.Dense(units=1, activation="sigmoid"))

# Check the structure of the model
nn_optimization3.summary()

Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense (Dense)               (None, 80)                2320      
                                                                 
 dense_1 (Dense)             (None, 30)                2430      
                                                                 
 dense_2 (Dense)             (None, 1)                 31        
                                                                 
Total params: 4781 (18.68 KB)
Trainable params: 4781 (18.68 KB)
Non-trainable params: 0 (0.00 Byte)
_________________________________________________________________


In [12]:
# Compile the model
nn_optimization3.compile(loss="binary_crossentropy", optimizer="adam", metrics=["accuracy"])

In [13]:
# Train the model
fit_model = nn_optimization3.fit(X_train_scaled,y_train,epochs=100)

Epoch 1/100
804/804 [==============================] - 5s 4ms/step - loss: 0.6098 - accuracy: 0.6932
Epoch 2/100
804/804 [==============================] - 3s 4ms/step - loss: 0.6044 - accuracy: 0.6967
Epoch 3/100
804/804 [==============================] - 2s 3ms/step - loss: 0.6024 - accuracy: 0.6977
Epoch 4/100
804/804 [==============================] - 2s 2ms/step - loss: 0.6015 - accuracy: 0.6981
Epoch 5/100
804/804 [==============================] - 2s 2ms/step - loss: 0.6014 - accuracy: 0.6983
Epoch 6/100
804/804 [==============================] - 2s 2ms/step - loss: 0.6007 - accuracy: 0.6987
Epoch 7/100
804/804 [==============================] - 2s 2ms/step - loss: 0.6001 - accuracy: 0.6990
Epoch 8/100
804/804 [==============================] - 2s 2ms/step - loss: 0.6002 - accuracy: 0.6993
Epoch 9/100
804/804 [==============================] - 3s 4ms/step - loss: 0.6004 - accuracy: 0.6993
Epoch 10/100
804/804 [==============================] - 2s 3ms/step - loss: 0.5998 - accura

In [14]:
# Evaluate the nn_optimization1 model using the test data
model_loss, model_accuracy = nn_optimization3.evaluate(X_test_scaled,y_test,verbose=2)
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")

268/268 - 1s - loss: 0.6022 - accuracy: 0.7010 - 527ms/epoch - 2ms/step
Loss: 0.6021933555603027, Accuracy: 0.7009912729263306


Original Model-
*   Loss: 0.559
*   Accuracy: 0.728

nn_optimization3-
*   Loss: 0.602
*   Accuracy: 0.701
*   Changes: the data being evaluated removed the CLASSIFICATION and APPLICATION_TYPE data fields
*   Model appears to be worse. Loss increased and Accuracy decreased.





  